### DISCUSSION TITLE
distinct CC# mapping table if tabindex=

### DISCUSSION THREAD

**Discussion by thethirdapple** [RomanKubiak/ctrlr#541] [Q&A] [answered]
A continuation of my first panel based on solutions provided in thread:540, thanks to #dnaldoog and #damiensellier.

My goal is to reuse/recycle a limited quantity of CC#'s by essentially enabling a new meta table based on if tabindex=
So that a distinct sliders on each tab could receive the same CC# from a BCR but send different Sysex to the synth...

Have been doing some [dr frankenstein] hacks by scraping from a few demos and og Ctrlr threads  but have not achieved life... yet. My script compiles:"ok" and executes when on tab=0 but not on tab=1 ??? still a noob at this...

I have over 407 Sysex messages, for RealTime midi modulation controls,  to send but only 32 knobs ;) per preset on the BCR. My Ctrlr panel currently receives CC# 127 from the BCR in +1 or -1 increments to switch thru tabs and returns a Sysex to the BCR to switch presets. So yes I can achieve 480 knobs but  limited to 127 CC's from the BCR.  I am attempting to send only a few direct CC messages and mainly Sysex to stay inline with Stephen Kay's programming of KARMA midi algorithms on the synth's and software he created, to avoid further "translation" headaches when integrating a final Ctrlr panel in the mix.

Peace

**Comment by dnaldoog** [comment]
Do you mean that instead of mapping one incoming CC message to one control, you want to map that incoming CC message to multiple controls on different tabs?

For example, you might have LFO1,LFO2,LFO3,LFO4  etc needing to send out 4 messages back to the Korg?

**Reply by thethirdapple** [reply]
Greetings,

Yes I think thats it.
hrrmmm... Yeah sorry for that hard to understand ramble.

incoming CC#70 assigned to slider1 on tab=1 sending  F0 something F7
incoming CC#70 assigned to slider2 on tab=2 sending F0 something else F7
incoming CC#70 assigned to slider3 on tab=3 sending F0 something else else F7
etc..

I thought, if possible, then the 407 sysex messages could be divided into +/-12 tabs, each tab with 32 sliders using a pool of only 32 cc messages coming from the external BCR controller...

or convince Uli to make this... 
![BCR-infinity](https://user-images.githubusercontent.com/125695472/221078196-495b9644-f78d-4de0-b400-3753be78be3f.png)

**Reply by dnaldoog** [reply]
So each bank of controls is the same ie `resonance1,resonance2,resonance3 ... 12` ?

**Reply by thethirdapple** [reply]
I was looking thru the Johnson Amp Panel you made and trying to do something similar. Where the drop down menu's use an "external" range table lookup. but use tabs instead of drop down menu...

[Ctrlr Panel.app.zip](https://github.com/RomanKubiak/ctrlr/files/10820397/Ctrlr.Panel.app.zip)

**Reply by thethirdapple** [reply]
I can kinda of figure out the " if tabindex=0 then enable "
but cannot figure out how to have more than one component name referenced in the midi meta table...

**Reply by thethirdapple** [reply]
To where t[k] is the variable looked up in the table would have more than one named slider referenced for any given incoming cc#... so I thought to do a map per tab.


myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local s = midi:getSize()
    if midi:getType() == 0 then -- it is a CC message
        local number = midi:getData():getByte(1)
        local value = midi:getData():getByte(2)
        _m(mapping[number]):setModulatorValue(value, false, true, false)
    end
end

meta = {}
meta.__index = function(t, k) -- fallback value in case a cc number comes in but there is no corresponding index in the table mapping
	_c.d:setText(string.format("CC number #%d not mapped!",k)) -- writes a value to a hidden modulator
    t[k] = "nullbin"
-- 		utils.warnWindow("Error", "number not mapped to a control") -- comment this out to die silently
    return t[k]
end


--
--KARMA FUNCTION MAPPING
--[cc#]="component name" -- value * 0-127, ** 0-63(off) 64-127(on), *** 0(off) 1-127(on) / KARMA description


mapping = {
[0]="bankselect",         -- *		MSB
[1]="mod1",                   -- *		Joystick +Y (up)
[2]="mod2",                  -- *		Joystick -Y (down)
[3]="",                            -- *		not assigned
[4]="footcontroller",     -- *		Foot Pedal function	
[5]="portamentoT",       -- *		Portamento Time
[6]="dataentry",            -- *		MSB of RPN and NRPN data
[7]="volume",	         -- * 	Volume
[8]="postfxpan",    	-- *		Pan post FX insert
[9]="",                             -- 			not assigned
[10]="pan",                     -- *		Panpot

etc...

**Comment by dnaldoog** [comment]
This panel works!

Code is changed and much more complex:

<pre>
myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local s = midi:getSize()
    if midi:getType() == 0 then -- it is a CC message
        local number = midi:getData():getByte(1)
        local value = midi:getData():getByte(2)
        local cycle = mapping[number].groupof or 0
        if cycle == 0 then
            return
        end
        for i = 1, cycle do
            local tmp_name = string.format("%s %d", mapping[number].name, i)
            _m(tmp_name):setModulatorValue(value, false, true, false)
        end
    end
end

meta = {}
meta.__index = function(t, k) -- fallback value 
    -- writes a value to a hidden modulator
    _c.d:setText(string.format("CC number #%d not mapped!", k))
    t[k] = "nullbin"
    --utils.warnWindow("Error", "number not mapped to a control") -- comment this out to die silently
    return t[k]
end

mapping = {
    [23] = {name = "resonance", groupof = 12},
    [36] = {name = "cutoff", groupof = 11}
}
setmetatable(mapping, meta)
</pre>

[map cc to sysex_3_0.zip](https://github.com/RomanKubiak/ctrlr/files/10820533/map.cc.to.sysex_3_0.zip)

**Reply by dnaldoog** [reply]
`groupof` is the number of repeated controls. In this case there are only 11 `cutoff`, but presumably there would usually be the exact same number of controls mirrored on each of the 12 (or so) tabs. So usually `groupof` would be 12 etc.

**Reply by dnaldoog** [reply]
This table -
<pre>
mapping = {
[0]="bankselect", -- * MSB
[1]="mod1", -- * Joystick +Y (up)
[2]="mod2", -- * Joystick -Y (down)
[3]="", -- * not assigned
[4]="footcontroller", -- * Foot Pedal function
[5]="portamentoT", -- * Portamento Time
[6]="dataentry", -- * MSB of RPN and NRPN data
[7]="volume", -- * Volume
[8]="postfxpan", -- * Pan post FX insert
[9]="", -- not assigned
[10]="pan", -- * Panpot
</pre>
becomes something like:
<pre>
mapping = {
    [0] = {name = "bankselect", groupof = 1}, -- * MSB
    [1] = {name = "mod", groupof = 2}, -- * Joystick +Y-Y (up)
    [4] = {name = "footcontroller", groupof = 1}, -- * Foot Pedal function
    [5] = {name = "portamentoT", groupof = 12}, -- * Portamento Time
    [6] = {name = "dataentry", groupof = 1}, -- * MSB of RPN and NRPN data
    [7] = {name = "volume", groupof = 12}, -- * Volume
    [8] = {name = "postfxpan", groupof = 12}, -- * Pan post FX insert
    [10] = {name = "pan", groupof = 12} -- * Panpot
}

</pre>

**Reply by thethirdapple** [reply]
Simply beautiful !!! Perhaps more complex code but a whole lot better than my messy round about attempts...
It will take a moment to understand what is happening. At first glance the, groupof= ,is indicating a limit to the cycle routine?

Would this need to be coded so that each slider on each tab has a different component label in this new map? 

It seems to me that conditional if tabindex= (enable true false) requires a component label to enable/disable. Or could it simple enable/disable the entire tab without reference to each component name?

such as this:
Or even better:
panel:getModulatorByName("somemodulator"):getComponent():setEnabled(false)

In the morning I will integrate this into my panel and continue m learning curve but I'm not exactly sure where to place the if tabindex= condition?

Sleep is what is needed to see this more clearly...
Again thank you for this super rapid path to a solution !!!

**Reply by dnaldoog** [reply]
Oh so do you mean that if a tab is visible, in other words the currently open tab, you only want that control sitting in that tab to change when the cc comes in?

**Reply by thethirdapple** [reply]
Yes, this is the idea...

**Comment by dnaldoog** [comment]
> So each bank of controls is the same ie resonance1,resonance2,resonance3 ... 12 ?

So the central design question is "*Do you want a single control on the external controller to change banks of controls with different names or will the names be all the same*?

1. Does say CC 23 change `resonance 1` ...**n** as well as `cutoff 1` ...**n**?
2. Does say CC 23 change `resonance 1` ...**n** but just `cutoff 1` and `mod1` and `lfo rate 1` ...n?
3. Does say CC 23 change `resonance 1` ...**n** _only_? (__the panel I just sent through__)

P.S. 1 and 2 will make things much much more complex. Remembering that increasing complexity will lead to increasing bugs!

To enable/disable or hide/show a component depending on the CC coming in would require a separate coding routine. Probably would call a function from the table mapping for such things.

**Comment by dnaldoog** [comment]
This panel only sends out sysex from the controls in currently selected tab only (from CC in). I think this is what you want?
This code is a modification of code from panel version 1.0 and 2.0 and so is not the same as version 3.0

<pre>
myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local currentTab = tonumber(_c.MYBIGTAB:getProperty("uiTabsCurrentTab") + 1) 
-- tabs start at zero but uiSliders' names in this panel start at 1 - resonance 1 - 12 etc in tabs 0-11
    local s = midi:getSize()
    if midi:getType() == 0 then -- it is a CC message
        local number = midi:getData():getByte(1)
        local value = midi:getData():getByte(2) -- a function is called to construct the name from uiCurrentTab
        local r = returnCcMappedControl(number, currentTab)
        _m(r):setModulatorValue(value, false, true, false)
    end
end

meta = {}
meta.__index = function(t, k) -- fallback value    -- writes a value to a hidden modulator
    _c.d:setText(string.format("CC number #%d not mapped!", k))
    t[k] = "nullbin"
    return t[k]
end

mapping = {
    [23] = "resonance",
    [36] = "cutoff"
}
setmetatable(mapping, meta)

returnCcMappedControl = function(num, tab)
    local name = mapping[num]
    if name == "nullbin" then
        return name
    else
        return string.format("%s %d", name, tab)
    end
end

</pre>

[map cc to sysex_4_0_version 2 altered.zip](https://github.com/RomanKubiak/ctrlr/files/10824346/map.cc.to.sysex_4_0_version.2.altered.zip)

**Reply by thethirdapple** [reply]
Will have a look in a second...

Truly appreciate the engagement and your patience to understand my ask. I am starting to see how Lua code functions, as I am a very visual learner, simply reading about it doesn't always translates to understanding. Being able to have demo's and other examples to "deconstruct" is a huge part of how I learn.

**Comment by dnaldoog** [comment, marked as answer]
I've just very quickly knocked this one together. So __not__ fully tested!

In this panel there is no `cutoff 12` but the panel doesn't crash. There are also single controls `bankselect` and `volume` so you can now do CC mapping for single controls as well as multiple controls with the same name in each tab, although as I said there is no `cutoff 12` in `tab 11`. If the table key `multi` is set to `false` the table key `q` is never accessed, but I just default it to `1`.

If you have a record in `mapping` but no corresponding Ctrlr object, the panel will crash.

<pre>
myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local currentTab = tonumber(_c.MYBIGTAB:getProperty("uiTabsCurrentTab") + 1) -- tabs start at zero but uiSliders start at 1
    local s = midi:getSize()
    if midi:getType() == 0 then -- it is a CC message
        local number = midi:getData():getByte(1)
        local value = midi:getData():getByte(2)

        local r = returnCcMappedControl(number, currentTab)
        _m(r):setModulatorValue(value, false, true, false)
    end
end

meta = {}
meta.__index = function(t, k) -- fallback value in case a cc number comes in but there is no corresponding index in the table mapping
    -- writes a value to a hidden modulator
    _c.d:setText(string.format("CC number #%d not mapped!", k))
    t[k] = "nullbin"
    return t[k]
end

mapping = {
    [0] = {n = "bankselect", multi = false, q = 1}, -- * MSB
    --[1] = {n = "mod", multi = false, q = 1}, -- * Joystick +Y-Y (up)
    --[4] = {n = "footcontroller", multi = false, q = 1}, -- * Foot Pedal function
    --[5] = {n = "portamentoT", multi = false, q = 1}, -- * Portamento Time
    --[6] = {n = "dataentry", multi = false, q = 1}, -- * MSB of RPN and NRPN data
    [7] = {n = "volume", multi = false, q = 1}, -- * Volume
    --[8] = {n = "postfxpan", multi = false, q = 1}, -- * Pan post FX insert
    --[10] = {n = "pan", multi = true, q = 12}, -- * Panpot
    [23] = {n = "resonance", multi = true, q = 12},
    [36] = {n = "cutoff", multi = true, q = 11}
}
setmetatable(mapping, meta)

returnCcMappedControl = function(num, tab)
    local name = mapping[num]
    if name == "nullbin" then
        return name
    elseif mapping[num].multi then
        if tab > mapping[num].q then
            return "nullbin"
        else
            return string.format("%s %d", name.n, tab)
        end
    else
        return name.n
    end
end

</pre>

[map cc to sysex_5_0_version 4 altered.zip](https://github.com/RomanKubiak/ctrlr/files/10824985/map.cc.to.sysex_5_0_version.4.altered.zip)

----

__EDIT:__
Actually the 'q' parameter in the table is not so useful and would only apply if the last `n` tabs had no records; in other words if a control is missing from any beginning tabs, the code won't work ― so not so useful!

**Reply by thethirdapple** [reply]
Testing v4 which works really well...

Tab0  "picksup" the adjustments from the other tabs. 
But tabs1-11 are completely independent !!!

having a look at v5 now...

**Reply by thethirdapple** [reply]
V5 is a work of a master! as are the previous versions as well :)

I can certainly live with simply "hiding" the sliders on tab0, as removing either tab0 and or the sliders on tab0 does crash the panel.

**Reply by thethirdapple** [reply]
[map cc to sysex.app.zip](https://github.com/RomanKubiak/ctrlr/files/10825431/map.cc.to.sysex.app.zip)

**Reply by dnaldoog** [reply]
Great! (Master = not!) But can you confirm that this panel - see attached - works with the Mac command key (I am sure it will but I just need to confirm) 

[Ctrl Key Down or not_2_0.zip](https://github.com/RomanKubiak/ctrlr/files/10829294/Ctrl.Key.Down.or.not_2_0.zip)

**Reply by thethirdapple** [reply]
In the land of the blind, the one-eyed man is king ! 

Will send over my test panel... have done some mod's to it prior to integrating into my project panel.

Ctrl Key Down: FUNCTIONS on Mac 10.13.6 running Ctrlr Version = 5.3.198, Build date = Tue Mar 15 15:53:07 PDT 2016, Branch = Nightly, Juce = 4.0.2, libusb = 1.0.19, liblo = 0.28

![Screen Shot 2023-02-24 at 6 34 36 PM](https://user-images.githubusercontent.com/125695472/221321189-01268c44-23e0-4c19-8a84-0aeca9a1e406.png)
![Screen Shot 2023-02-24 at 6 34 46 PM](https://user-images.githubusercontent.com/125695472/221321197-ca31c424-02eb-47cc-b818-a4cb9d8dbd06.png)

**Comment by dnaldoog** [comment]
In this version I tackle the problem of having separate controls in each tab. With this code you can put  controls in any tab in any order as long as the name of the control suffix matches the `uiTabIndex` number:

<pre>
myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local currentTab = tonumber(_c.MYBIGTAB:getProperty("uiTabsCurrentTab"))
    local s = midi:getSize()
    if midi:getType() == 0 then -- it is a CC message
        local number = midi:getData():getByte(1)
        local value = midi:getData():getByte(2)

        local r = returnCcMappedControl(number, currentTab)
        _m(r):setModulatorValue(value, false, true, false)
    end
end

meta = {}
meta.__index = function(t, k) -- fallback value in case a cc number comes in but there is no corresponding index in the table mapping
    -- writes a value to a hidden modulator
    _c.d:setText(string.format("CC number #%d not mapped!", k))
    t[k] = "nullbin"
    return t[k]
end

mapping = {
    [0] = {n = "bankselect", multi = false, q = {nil}}, -- * MSB
    [7] = {n = "volume", multi = false, q = {nil}}, -- * Volume
    [21] = {n = "lfo delay", multi = true, q = {0}},
    [23] = {n = "resonance", multi = true, q = {0,5,7,9,11}},
    [36] = {n = "cutoff", multi = true, q = {0,1,2}}
}
setmetatable(mapping, meta)

returnCcMappedControl = function(num, tab)
    local name = mapping[num]
    if name == "nullbin" then
        return name
    elseif mapping[num].multi then
        local found = false
        for i, v in ipairs(mapping[num].q) do
            if v == tab then
                found = true
                return string.format("%s %d", name.n, tab)
            end
        end
        if found == false then
            return "nullbin"
        end
    else
        return name.n
    end
end
</pre>

[map cc to sysex_7_0.zip](https://github.com/RomanKubiak/ctrlr/files/10830713/map.cc.to.sysex_7_0.zip)

**Reply by thethirdapple** [reply]
This is fantastic so as to avoid panel clutter !!!

**Comment by dobo365** [comment]
Just reading quickly in diagonal this long thread... didn't look in detail to the solution
So, simply my 2c ;-)

I would have done the following:
- in all tabs, just assign the CC's as they should
- set all dynamic modulators on Mute
- when clicking a tab, unMute all the modulators with a CC property (only dynamic modulators) with Tab Index= current_tab

This is requiring just a few lines of code...

**Reply by dnaldoog** [reply]
How do you mute a modulator?

**Reply by thethirdapple** [reply]
here is a diagram of my panel functions and why V5_4 was marked as a solution: the panel sliders only react to the BCR input based on which tab is in focus... there might well be a more efficient method but works for now...

![KARMA function layout](https://user-images.githubusercontent.com/125695472/221382306-8fbb55b8-eb4c-43e1-b019-6657f484e987.jpg)

**Reply by damiensellier** [reply]
Hi guys, you've done a great job! 
You made me think I was working on a "midi learn" function for Ctrlr. For this kind of panels with a lot of CC controllers this is where it would be awesome. 
Perhaps just assigning your different tabs to MIDI Channels would help with the crazy amount of CC you have with this pannel like One BCR for each channel assigned to 1 Tab. Can you change easily the Output Midi Channel easily from the BCR? 1 channel would equal to 1 BCR for 1 Ctrlr Tab. If your Inc/Dec buttons can shift Midi Output Channel that would be more efficient from what I understand.

Edit : I also recommand coding with the latest macOS version avalaible on godslikeproduction website : https://godlike.com.au/index.php?id=361 it's on the right sidebar. it's v5.6.23 and the problem you faced about loading and assigning images for modulators with the greyed out drop-down menu is fixed up as many other bugs.

**Reply by thethirdapple** [reply]
The BCR/BCF are not super tedious to program... There is a single global Midi channel available, but have to be careful to avoid midi loops as the boxes do allow merge from the "in" to the separate dedicated "out and thru". 

And each knob/slider(fader)/button can have a distinct midi channel assigned to it which is nice. 

Behringer announced, but not released updated version. I have not seen it nor read too much about it other than the press release... looks like they integrated the ZAQuencer mod thats been available for while...

https://macprovideo.com/article/audio-hardware/behringer-reveals-features-for-upcoming-bcr32-usb-midi-controller-sequencer

**Reply by dnaldoog** [reply]
How does it handle negative values?

**Reply by thethirdapple** [reply]
![tab_preset hack](https://user-images.githubusercontent.com/125695472/221384773-4526ee50-a2b5-439e-9f12-7fb25d0a12a5.jpg)

“If you don’t know where you are going any road can take you there.”
- Cheshire Cat / Alice’s Adventures in Wonderland, Lewis Carrol

[map cc to sysex_5_3.panel.zip](https://github.com/RomanKubiak/ctrlr/files/10832556/map.cc.to.sysex_5_3.panel.zip)

**Reply by damiensellier** [reply]
> The BCR/BCF are not super tedious to program... There is a single global Midi channel available, but have to be careful to avoid midi loops as the boxes do allow merge from the "in" to the separate dedicated "out and thru".
> 
> And each knob/slider(fader)/button can have a distinct midi channel assigned to it which is nice.
> 
> Behringer announced, but not released updated version. I have not seen it nor read too much about it other than the press release... looks like they integrated the ZAQuencer mod thats been available for while...
> 
> https://macprovideo.com/article/audio-hardware/behringer-reveals-features-for-upcoming-bcr32-usb-midi-controller-sequencer

yeah I know the announce BCR/BCF controllers look so much better and will probably have MCU integration included. Like every Behringer gear, they announce it, but maybe it won't ever be released. That's their problem. My main controller is the Midi Fighter Twister with 4x4knobs x4 banks it's my favorite CC controller for the moment.

**Reply by dnaldoog** [reply]
This version fixes some crashes  in _map cc to sysex_5_3.panel.zip_ and handles negative sysex mapped from positive CC 0-127

<pre>
myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
    local currentTab = tonumber(_c.tabs:getProperty("uiTabsCurrentTab") + 1) -- tabs start at zero but uiSliders start at 1
    local s = midi:getSize()
    if midi:getType() == 0 then -- it is a CC message
        local number = midi:getData():getByte(1)
        local value = midi:getData():getByte(2)

        local r, val = returnCcMappedControl(number, currentTab, value)
        _m(r):setModulatorValue(val, false, true, false)
    end
end

meta = {}
meta.__index = function(t, k) -- fallback value in case a cc number comes in but there is no corresponding index in the table mapping
    console(string.format("CC number #%d not mapped!", k))
    -- writes a value to a hidden modulator

    t[k] = {n = "nullbin", multi = false, min = 0}
    return t[k]
end

mapping = {
    [7] = {n = "volume", multi = false, q = 1, min = 0},
    [23] = {n = "knobA", multi = true, q = 3, min = 0},
    [24] = {n = "knobB", multi = true, q = 3, min = 0},
    [25] = {n = "knobC", multi = true, q = 3, min = 0},
    [26] = {n = "knobD", multi = true, q = 3, min = 0},
    [27] = {n = "knobE", multi = true, q = 3, min = 0},
    [28] = {n = "knobF", multi = true, q = 3, min = 0},
    [29] = {n = "knobG", multi = true, q = 3, min = 50}, --50/+50
    [30] = {n = "knobH", multi = true, q = 3, min = 0},
    [31] = {n = "knobI", multi = true, q = 3, min = 0},
    [127] = {n = "BCRpresets", multi = false, q = 1, min = 0}
}

setmetatable(mapping, meta)

returnCcMappedControl = function(num, tab, value)
    local name = mapping[num].n
    local valmap = mapping[num].min or 0
    local value = value - valmap
    local ismulti = mapping[num].multi
    local q = mapping[num].q
    if name == "nullbin" then
        return name, value
    elseif ismulti then
        if tab > q then
            return "nullbin", value
        else
            return string.format("%s %d", name, tab), value
        end
    else
        return name, value
    end
end

</pre>

[map cc to sysex_5_4.zip](https://github.com/RomanKubiak/ctrlr/files/10833275/map.cc.to.sysex_5_4.zip)

**Reply by thethirdapple** [reply]
Ah right ok, misunderstood your question... 

Pre-emptive answers to questions not yet posed! love it!

Again and always thank you to everyone chiming into this long, forked, thread.

**Reply by dobo365** [reply]
Sorry for delayed answer...

I wrote a function to mute/unmute all modulators. You need to check for dynamic modulators only.
```
function muteModulators(bMute)

	n = panel:getNumModulators()

	if bMute then
		iMute = 1
	else
		iMute = 0
	end
		
	for i = 0, n-1 do

		local mod = panel:getModulatorByIndex(i)
--		local Name= L(mod:getName())

		if mod:getProperty("modulatorIsStatic") == "0" then --	 Only Dynamic modulators have a MidiMessage ;-)
			-- console(tostring(i.." "..Name))
			mod:setPropertyInt("modulatorMute", iMute)
		end
	end

end
```


Now, for your case, you have 2 ways for doing that:
Option1: you keep the method above for muting/unmuting all tabs and you add a copy of the method by adding an extra variable being the tab you want to unmute and you add an extra check for the TabIndex of the modulator being looked at. Only unmute if match.
Lua is fast so even with a lot of modulators it will still work fine.
Option2: you create a table for each tab. Each table will contain the list of dynamic modulators on that tab. you keep the method above for muting/unmuting all tabs and you add a copy of the method by adding an extra variable being the table you want to unmute and you directly mute/unmute all modulators of the given table. Much more efficient and fast. The drawback is that you need to maintain the different tables.

For both methods, you use them in a method Tab_OnChange (on uiTabs, called when current tab is changed):
```
--
-- Called when the current tab in an uiTabs component is changed
--
Tab_OnChange = function(--[[ CtrlrModulator --]] mod, --[[ number --]] tabIndex)

	-- No action while loading a program or if the panel is in bootstrap or program states
	if not isPanelReady() or bPanelLoaded == false or bPanelLoaded == nil then
		return
	end

	muteModulators(false, tabIndex)

end

```

**Reply by thethirdapple** [reply]
This is truly expanded my horizons… thank you. Will look at this, in detail, tomorrow and try it to mix this script with other solutions. Already see some interesting usages I had not thought of previously. 

Thank you
